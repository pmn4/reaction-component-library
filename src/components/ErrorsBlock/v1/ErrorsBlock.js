import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import isEmpty from "lodash.isempty";
import { applyTheme } from "helpers";

const ErrorWrapper = styled.div`
  margin: ${applyTheme("errorsBlockMargin")};
`;

const Error = styled.div`
  color: ${applyTheme("errorsBlockColor")};
  font-size: ${applyTheme("errorsBlockFontSize")};
`;

const IconWrapper = styled.div`
  color: ${applyTheme("errorsBlockColor")};
  display: inline-block;
  fill: currentColor;
  margin: ${applyTheme("errorsBlockIconMargin")};

  & * {
   color: inherit;
   fill: inherit;
  }
`;

class ErrorsBlock extends Component {
  static isFormErrors = true;

  static propTypes = {
    className: PropTypes.string,
    errorClassName: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    errors: PropTypes.array,
    iconError: PropTypes.node,
    names: PropTypes.arrayOf(PropTypes.string),
    shouldShowIcon: PropTypes.bool
  };

  static defaultProps = {
    className: undefined,
    errorClassName: undefined,
    errors: undefined,
    iconError: (<i className="fas fa-exclamation-triangle" />),
    names: undefined,
    shouldShowIcon: false
  };

  renderIcon() {
    const { iconError } = this.props;
    return <IconWrapper>{iconError}</IconWrapper>;
  }

  render() {
    const { errorClassName, errors, className, shouldShowIcon } = this.props;

    if (isEmpty(errors)) return null;

    // https://reactjs.org/docs/lists-and-keys.html
    // "When you don’t have stable IDs for rendered items, you may use the item index as a key as a last resort.
    // We don’t recommend using indexes for keys if the items can reorder, as that would be slow."
    //
    // There will rarely be more than a few errors for a field, and forcing unique ID
    // generation for them would be arbitrary and unnecessary. So we'll use the index.
    return (
      <ErrorWrapper className={className}>
        {errors.map((error, index) => (
          <Error key={index} className={errorClassName} data-name={error.name}>
            {shouldShowIcon ? this.renderIcon() : ""}
            {error.message}
          </Error>
        ))}
      </ErrorWrapper>
    );
  }
}

export default ErrorsBlock;
