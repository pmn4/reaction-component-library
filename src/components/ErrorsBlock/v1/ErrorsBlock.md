### ErrorsBlock Overview

```jsx
const errors = [{ name: "example", message: "This field is required" }];
<Field name="example" label="Label" errors={errors}>
 <TextInput name="example" errors={errors} />
 <ErrorsBlock names={["example"]} errors={errors} />
</Field>
```

```jsx
const errors = [{ name: "example", message: "This field is required" }, { name: "example", message: "Another error" }];
<Field name="example" label="Label" errors={errors}>
 <TextInput name="example" errors={errors} shouldAllowLineBreaks />
 <ErrorsBlock names={["example"]} errors={errors} shouldShowIcon />
</Field>
```
