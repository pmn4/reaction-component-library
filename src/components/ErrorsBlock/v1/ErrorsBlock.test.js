import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import ErrorsBlock from "./ErrorsBlock";

test("basic snapshot", () => {
  const component = renderer.create(<ErrorsBlock />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
