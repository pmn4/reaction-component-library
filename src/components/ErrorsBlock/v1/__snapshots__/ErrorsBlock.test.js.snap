// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all error messages 1`] = `
.c0 {
  margin: 0.625rem 0 0;
}

.c1 {
  color: #cd3f4c;
  font-size: 0.875rem;
}

<div
  className="c0"
>
  <div
    className="c1"
    data-name="a"
  >
    
    Message One
  </div>
  <div
    className="c1"
    data-name="b"
  >
    
    Message Two
  </div>
</div>
`;

exports[`renders nothing when there are no errors 1`] = `null`;
